libs/icu_headroom_metric.py

```diff
@@ -0,0 +1,168 @@
+            return None
+
+        return actuals
+
+    @property
+    def total_icu_beds(self):
+        timeseries = self.data[CommonFields.ICU_BEDS]
+        if timeseries.any():
+            return timeseries
+        return self.latest_values[CommonFields.ICU_BEDS]
+
+    @property
+    def typical_usage_rate(self):
+        return self.latest_values[CommonFields.ICU_TYPICAL_OCCUPANCY_RATE]
+
+    @property
+    def current_icu_non_covid_with_source(self):
+        if self.actual_current_icu_covid is not None and self.actual_current_icu_total is not None:
+            source = NonCovidPatientsMethod.ACTUAL
+            return (self.actual_current_icu_total - self.actual_current_icu_covid, source)
+
+        if self.actual_current_icu_total is not None:
+            source = NonCovidPatientsMethod.ESTIMATED_FROM_TOTAL_ICU_ACTUAL
+            return (self.actual_current_icu_total - self.estimated_current_icu_covid, source)
+
+        # Do decomp
+        non_covid_utilization = self.typical_usage_rate - self.decomp_factor
+        source = NonCovidPatientsMethod.ESTIMATED_FROM_TYPICAL_UTILIZATION
+
+        if isinstance(self.total_icu_beds, pd.Series):
+            decomp = non_covid_utilization * self.total_icu_beds
+            decomp.name = None
+            return (decomp, source)
+
+        total_icu_beds = pd.Series([self.total_icu_beds] * len(self.data), index=self.data.index)
+        decomp = non_covid_utilization * total_icu_beds
+
+        return (decomp, source)
+
+    @property
+    def current_icu_covid_with_source(self):
+        if self.actual_current_icu_covid is not None:
+            source = CovidPatientsMethod.ACTUAL
+            return self.actual_current_icu_covid, source
+
+        source = CovidPatientsMethod.ESTIMATED
+        return self.estimated_current_icu_covid, source
+
+
+def calculate_icu_utilization_metric(icu_data: ICUMetricData):
+    """
+
+             covid icu patients
+           ----------------------
+    available beds for covid icu patients
+
+
+    If we know exactly how many beds are in use by non covid patients we can
+    calculate the number of available beds exactly.
+
+    otherwise available beds is calulated by looking at historical data and
+    assuming that we can add more capacity.
+
+    So, we need to know or guess how many icu beds are occupied by non covid patients
+    non covid patients:
+     - get from actuals
+     - estimate from typical utilization + decomp factor
+
+    available beds can be calulated
+
+                  covid icu patients
+                ----------------------
+           total beds - non covid patients
+                          ==
+                  covid icu patients
+                ---------------------
+    total beds - (typical non covid usage - decomp)
+                          or
+                  covid icu patients
+                ---------------------
+    total beds - (total covid patients - covid patients)
```

Comment:
"total covid patients - covid patients" seems wrong.

In general, I'm not sure how helpful this comment is (trying to describe all the estimation calculations up front instead of next to the code that implements it)

Reply:
ah yes `s/total covid/total icu/`

sorry I should've given a bit more of a heads up about this comment as I'm not sure if it's really that useful. it was more useful to me as I was working through the different options available.  I do think some context is useful, so i'll try to boil it down in a more approachable/concise manner